Analysis/Tasks/PWGHF/HFCandidateCreatorX.cxx

```diff
@@ -0,0 +1,300 @@
+  OutputObj<TH1F> hMassX{TH1F("hMassX", "3-prong candidates;inv. mass (J/#psi #pi+ #pi-) (GeV/#it{c}^{2});entries", 500, 0., 5.)};
+  OutputObj<TH1F> hCovPVXX{TH1F("hCovPVXX", "3-prong candidates;XX element of cov. matrix of prim. vtx position (cm^{2});entries", 100, 0., 1.e-4)};
+  OutputObj<TH1F> hCovSVXX{TH1F("hCovSVXX", "3-prong candidates;XX element of cov. matrix of sec. vtx position (cm^{2});entries", 100, 0., 0.2)};
+
+  double massPi = RecoDecay::getMassPDG(kPiPlus);
+  double massJpsi = RecoDecay::getMassPDG(443);
+  double massJpsiPiPi;
+
+  Configurable<int> d_selectionFlagJpsi{"d_selectionFlagJpsi", 1, "Selection Flag for Jpsi"};
+  Configurable<double> cutYCandMax{"cutYCandMax", -1., "max. cand. rapidity"};
+  Filter filterSelectCandidates = (aod::hf_selcandidate_jpsi::isSelJpsiToEE >= d_selectionFlagJpsi);
+
+  void process(aod::Collision const& collision,
+               soa::Filtered<soa::Join<
+                 aod::HfCandProng2,
+                 aod::HFSelJpsiToEECandidate>> const& jpsiCands,
+               aod::BigTracks const& tracks)
+  {
+    // 2-prong vertex fitter (to rebuild Jpsi vertex)
+    o2::vertexing::DCAFitterN<2> df2;
+    df2.setBz(magneticField);
+    df2.setPropagateToPCA(b_propdca);
+    df2.setMaxR(d_maxr);
+    df2.setMaxDZIni(d_maxdzini);
+    df2.setMinParamChange(d_minparamchange);
+    df2.setMinRelChi2Change(d_minrelchi2change);
+    df2.setUseAbsDCA(true);
+
+    // 3-prong vertex fitter
+    o2::vertexing::DCAFitterN<3> df3;
+    df3.setBz(magneticField);
+    df3.setPropagateToPCA(b_propdca);
+    df3.setMaxR(d_maxr);
+    df3.setMaxDZIni(d_maxdzini);
+    df3.setMinParamChange(d_minparamchange);
+    df3.setMinRelChi2Change(d_minrelchi2change);
+    df3.setUseAbsDCA(true);
+
+    // loop over Jpsi candidates
+    for (auto& jpsiCand : jpsiCands) {
+      if (!(jpsiCand.hfflag() & 1 << JpsiToEE)) {
+        continue;
+      }
+      if (cutYCandMax >= 0. && std::abs(YJpsi(jpsiCand)) > cutYCandMax) {
+        continue;
+      }
+      hMassJpsi->Fill(InvMassJpsiToEE(jpsiCand));
+      hPtJpsi->Fill(jpsiCand.pt());
+      hCPAJpsi->Fill(jpsiCand.cpa());
+      // create Jpsi track to pass to DCA fitter; use cand table + rebuild vertex
+      const std::array<float, 3> vertexJpsi = {jpsiCand.xSecondaryVertex(), jpsiCand.ySecondaryVertex(), jpsiCand.zSecondaryVertex()};
+      array<float, 3> pvecJpsi = {jpsiCand.px(), jpsiCand.py(), jpsiCand.pz()};
+      auto prong0 = jpsiCand.index0_as<aod::BigTracks>();
+      auto prong1 = jpsiCand.index1_as<aod::BigTracks>();
+      auto prong0TrackParCov = getTrackParCov(prong0);
+      auto prong1TrackParCov = getTrackParCov(prong1);
+
+      if (df2.process(prong0TrackParCov, prong1TrackParCov) == 0) {
+        continue;
+      }
+
+      // propogate prong tracks to Jpsi vertex
+      prong0TrackParCov.propagateTo(jpsiCand.xSecondaryVertex(), magneticField);
```

Comment:
I am not sure we don't need here the DCA. If it is a nonprompt X e.g. the Jpsi will also be displaced (I will think about it).